app/controllers/programming_groups_controller.rb

```diff
@@ -4,7 +4,14 @@ class ProgrammingGroupsController < ApplicationController
   include CommonBehavior
   include LtiHelper
 
-  before_action :set_exercise_and_authorize
+  before_action :set_exercise_and_authorize, only: %i[new create]
+  before_action :set_programming_group, only: %i[destroy]
+
+  def index
+    @search = ProgrammingGroup.ransack(params[:q], {auth_object: current_user})
+    @programming_groups = @search.result.includes(:exercise).order(:id).paginate(page: params[:page], per_page: per_page_param)
```

Comment:
This includes is not sufficient, we are still performing an N+1 query for the programming group memberships and users...

Let's go with:

```rb
includes(:exercise, :programming_group_memberships, :internal_users, :external_users)
```

Please take a moment to compare the SQL queries generated by your code and my suggestion to see the difference 🔍